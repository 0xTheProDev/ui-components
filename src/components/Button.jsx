// @flow
import React from 'react';
import cx from 'classnames';

type ButtonType = 'primary' | 'secondary' | 'danger';

export type ButtonProps = {
  children?: string,
  disabled?: boolean,
  loading?: boolean,
  badge?: ?number,
  small?: boolean,
  onDark?: boolean,
  type: ButtonType,
  onClick: Function,
};

const Button = ({
  children,
  type,
  badge,
  disabled,
  loading,
  small,
  onDark,
  onClick,
}: ButtonProps) => {
  const hasBadge: boolean = !!badge || badge === 0;
  return (
    <button
      className={cx('btn', `btn-${type}`, {
        'is-disabled': disabled,
        'is-loading': loading,
        'has-badge': hasBadge,
        'btn-small': small,
        'btn-on-dark': onDark,
      })}
      onClick={onClick}
    >
      {hasBadge && <span className="badge">{badge}</span>}
      {children}
    </button>
  );
};

Button.defaultProps = {
  children: 'Button Text',
  disabled: false,
  loading: false,
  small: false,
  onDark: false,
  badge: null,
};

export default Button;
