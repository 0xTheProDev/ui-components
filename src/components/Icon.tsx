import * as React from 'react';
import { IconType } from '../types/icons';

export type IconProps = {
  type: IconType;
};

const Icon: React.SFC<IconProps> = ({ type }) => (
  <i className={`sg-icon sg-icon-${type}`} />
);

export default Icon;
