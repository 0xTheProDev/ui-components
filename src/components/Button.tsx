import React from 'react';
import { IconType } from '../types/icons';
import cn from '../utilities/classnames';
import Badge from './Badge';
import Icon from './Icon';
import Loader from './Loader';

export type ButtonType = 'primary' | 'secondary' | 'danger' | 'group-item';

export interface AllButtonProps {
  disabled?: boolean;
  loading?: boolean;
  badge?: number | string;
  small?: boolean;
  onDark?: boolean;
  type?: ButtonType;
  onClick?: (event: any) => void;
  group?: boolean;
  active?: boolean;
  icon?: IconType;
  id?: string;
  isLink?: boolean;
}

export interface ButtonProps extends AllButtonProps {
  children?: string | React.ReactNode;
}

export interface ButtonizedProps extends AllButtonProps {
  children?: React.ReactElement<ButtonProps>;
}

export const Button: React.SFC<ButtonProps> = (props) => {
  return (
    <Buttonized {...props} >
      <button>
        {props.children}
      </button>
    </Buttonized>
  );
};

export const Buttonized: React.SFC<ButtonizedProps> = ({
  children,
  type,
  badge,
  disabled,
  loading,
  small,
  onDark,
  onClick,
  group,
  active,
  icon,
  id,
}) => {
  const hasBadge: boolean = !!badge || badge === 0;
  const hasIcon: boolean = !!icon;
  const content: any[] = [];

  if (hasBadge) {
    content.push(
      <Badge key={1}>{badge}</Badge>,
    );
  }

  if (hasIcon) {
    content.push(
      <Icon key={2} type={icon} />,
    );
  }

  if (loading) {
    content.push(
      <Loader key={3} small onDark={type === 'primary'} />,
    );
  }

  // the children of the element being buttonized
  if (children.props.children) {
    content.push(
      children.props.children,
    );
  }

  return (
    React.cloneElement(
      children,
      {
        className: cn('btn', `btn-${type}`, {
          'btn-on-dark': onDark,
          'btn-small': small,
          'has-badge': hasBadge,
          'has-icon': hasIcon || loading,
          'is-active': active,
          'is-disabled': disabled,
          'is-loading': loading,
        }),
        id,
        onClick,
      },
      content,
    )
  );
};

Button.defaultProps = {
  active: false,
  badge: null,
  children: '',
  disabled: false,
  group: false,
  icon: null,
  loading: false,
  onClick: () => {},
  onDark: false,
  small: false,
  type: 'primary',
};

export default Button;
