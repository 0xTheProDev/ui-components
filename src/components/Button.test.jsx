// @flow
import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';
import Icon from './Icon';

describe('Button', () => {
  const noop = () => { };

  it('should have a class of .btn', () => {
    const wrapper = shallow(
      <Button onClick={noop} type="secondary">Button</Button>,
    );
    expect(wrapper.hasClass('btn')).toBe(true);
  });

  it('should display the text passed in as a child', () => {
    const wrapper = shallow(
      <Button onClick={noop} type="secondary">Find Me</Button>,
    );
    expect(wrapper.text()).toBe('Find Me');
  });

  describe('Button.small', () => {
    it('should have a class of .btn-small if passed a property of "small"', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary" small>Button</Button>,
      );
      expect(wrapper.hasClass('btn-small')).toBe(true);
    });
  });

  describe('Button.onDark', () => {
    it('should have a class of .btn-on-dark if passed a property of "onDark"', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary" onDark>Button</Button>,
      );
      expect(wrapper.hasClass('btn-on-dark')).toBe(true);
    });
  });

  describe('Button.disabled', () => {
    it('should have a class of .is-disabled if passed a property of "disabled"', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary" disabled>Button</Button>,
      );
      expect(wrapper.hasClass('is-disabled')).toBe(true);
    });
  });

  describe('Button.loading', () => {
    it('should have a class of .is-loading if passed a property of "loading"', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary" loading>Button</Button>,
      );
      expect(wrapper.hasClass('is-loading')).toBe(true);
    });
  });

  describe('Button.badge', () => {
    it('should have a class of .has-badge is a badge count is provided', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary" badge={2}>Button</Button>,
      );
      expect(wrapper.hasClass('has-badge')).toBe(true);
    });

    it('should not have a class of .has-badge is a badge count is not provided', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary">Button</Button>,
      );
      expect(wrapper.hasClass('has-badge')).toBe(false);
    });

    it('should have a span with the class of .badge if a badge count is provided', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary" badge={2}>Button</Button>,
      );
      expect(wrapper.find('.badge').length).toBe(1);
    });

    it('should not have a span with the class of .badge if a badge count is not provided', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary">Button</Button>,
      );
      expect(wrapper.find('.badge').length).toBe(0);
    });

    it('should display the badge count in the .badge span', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary" badge={2}>Button</Button>,
      );
      expect(wrapper.find('.badge').text()).toBe('2');
    });
  });

  describe('Button types', () => {
    it('should have a class of .btn-primary if passed a type of "primary"', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="primary">Button</Button>,
      );
      expect(wrapper.hasClass('btn-primary')).toBe(true);
    });

    it('should have a class of .btn-secondary if passed a type of "secondary"', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="secondary">Button</Button>,
      );
      expect(wrapper.hasClass('btn-secondary')).toBe(true);
    });

    it('should have a class of .btn-primary if passed a type of "danger"', () => {
      const wrapper = shallow(
        <Button onClick={noop} type="danger">Button</Button>,
      );
      expect(wrapper.hasClass('btn-danger')).toBe(true);
    });
  });

  describe('Buttons with icons', () => {
    it('should match the snapshot', () => {
      const wrapper = shallow(
        <Button>
          <Icon type="x" /> Button
        </Button>,
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Button.onClick', () => {
    it('should fire the function passed in as "onClick" when it is clicked', () => {
      const clickAction = jest.fn();
      const wrapper = shallow(
        <Button onClick={clickAction} type="danger">Button</Button>,
      );

      wrapper.simulate('click');

      expect(clickAction.mock.calls.length).toBe(1);
    });
  });
});
