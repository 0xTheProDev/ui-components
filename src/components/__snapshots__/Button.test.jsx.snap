// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Buttons with icons should match the snapshot 1`] = `
<button
  className="btn btn-undefined"
>
  <Icon
    size="small"
    type="x"
  />
   Button
</button>
`;
